File: Hangman.py
import random

def get_word():
    # computer randomly choose a word that is going to be guessed from the word list
    word_list = ['red', 'orange', 'yellow', 'green', 'blue', 'indigo', 'purple']
    word = random.choice(word_list)
    return word.upper()

def play(word):

    # initialize variables
    word_completion = "_" * len(word)
    guessed = False
    guessed_letters = []
    tries = 6

    # first printing
    print("Let's play Hangman!")
    print(display_hangman(tries))
    print("Your word (", len(word), "letters ):", word_completion)
    print("\n")

    # loop
    while not guessed and tries > 0:

        # get player's guess
        guess = input("Please guess a letter: ").upper()

        # check the input guess whether is an alphabet letter
        # len(guess) is the length of the input guess.
        # isalpha() to check whether the input guess contains alphabet only.
        if len(guess) == 1 and guess.isalpha():

            # if the input guess is already guessed
            if guess in guessed_letters:
                print("You already guessed the letter", (guess + ". Please try again."))

            # if the input guess is wrong, is not in the WORD
            elif guess not in word:
                print(guess, "is not in the word.")
                # deduct the tries by one
                tries -= 1
                # add the guessed letter to the guessed_letters list
                guessed_letters.append(guess)

            # if the input guess is correct, is in the WORD
            else:
                print("Good job,", guess, "is in the word!")
                # add the guessed letter to the guessed_letters list
                guessed_letters.append(guess)
                # replace the respective unknown symbol with the correct letter
                word_as_list = list(word_completion)
                indices = [i for i, letter in enumerate(word) if letter == guess]
                for index in indices:
                    word_as_list[index] = guess
                word_completion = "".join(word_as_list)

                # check if there is still have any unknown symbol in the guessing WORD
                if "_" not in word_completion:
                    guessed = True

        elif len(guess) == len(word) and guess.isalpha():
            if guess in guessed_letters:
                print("You already guessed the word", (guess + "."))
            elif guess != word:
                print(guess, "is not the word.")
                tries -= 1
                guessed_letters.append(guess)
            else:
                guessed = True
                word_completion = word

        # if the input guess is not an alphabet letter or the input guess is not a valid word guess
        else:
            print("Not a valid guess. Please try again.")

        # print current state of the hangman and the progress of the guessing WORD
        print(display_hangman(tries))
        print("Your word (", len(word), "letters ):", word_completion)
        print("\n")

    # final result
    # if the WORD is guessed correctly
    if guessed:
        print("Congratulations, you guessed the word! You win!")
    # if the player has run out of tries
    else:
        print("Sorry, you ran out of tries. The word was " + word + ". Maybe next time!")


def display_hangman(tries):
    stages = [  # final state: head, torso, both arms, and both legs
                """
                   --------
                   |      |
                   |      O
                   |     \\|/
                   |      |
                   |     / \\
                   -
                """,
                # head, torso, both arms, and one leg
                """
                   --------
                   |      |
                   |      O
                   |     \\|/
                   |      |
                   |     / 
                   -
                """,
                # head, torso, and both arms
                """
                   --------
                   |      |
                   |      O
                   |     \\|/
                   |      |
                   |      
                   -
                """,
                # head, torso, and one arm
                """
                   --------
                   |      |
                   |      O
                   |     \\|
                   |      |
                   |     
                   -
                """,
                # head and torso
                """
                   --------
                   |      |
                   |      O
                   |      |
                   |      |
                   |     
                   -
                """,
                # head
                """
                   --------
                   |      |
                   |      O
                   |    
                   |      
                   |     
                   -
                """,
                # initial empty state
                """
                   --------
                   |      |
                   |      
                   |    
                   |      
                   |     
                   -
                """
    ]
    return stages[tries]

def main():
    word = get_word()
    play(word)
    while input("Play Again? (Y/N) ").upper() == "Y":
        word = get_word()
        play(word)

main()

File: Hangman.java
import java.util.ArrayList;
import java.util.Random;
import java.util.Scanner;

public class Hangman {

	private static final String ANSI_RED = "\033[0;31m";
	private static final String ANSI_GREEN = "\033[0;32m";
	private static final String ANSI_CYAN = "\033[0;36m";
	private static final String ANSI_RESET = "\u001B[0m";

	public static String get_word() {

		Random rand = new Random();
		String[] word_list = { "red", "orange", "yellow", "green", "blue", "indigo", "purple" };
		String word = word_list[rand.nextInt(word_list.length)];
		word = word.toUpperCase();
		return word;

	}

	public static void play(String word) {

		Scanner scan = new Scanner(System.in);

		String symbol = "*";
		String word_completion = new String(new char[word.length()]).replace("\0", symbol);
		boolean guessed = false;
		ArrayList<String> guessed_letters = new ArrayList<String>();
		int tries = 6;

		System.out.println("Let's play Hangman!");
		display_Hangman(tries);
		System.out.println("Your word (" + word.length() + " letters): " + word_completion);
		System.out.println();

		while (!guessed && tries > 0) {
			System.out.print("Please type your guess: ");
			String guess = scan.nextLine();
			guess = guess.toUpperCase();

			if (guess.length() == 1 && guess.matches("[a-zA-Z]+")) {
				if (guessed_letters.contains(guess)) {
					System.out.println("You already guessed the letter " + guess + ". Please try again!");
				} else if (!(word.contains(guess))) {
					System.out.println(guess + " is not in the WORD. Please try again!");
					tries = tries - 1;
					guessed_letters.add(guess);
				} else {
					System.out.println("Good Job! " + guess + " is in the WORD.");
					guessed_letters.add(guess);
					char[] ch1 = word.toCharArray();
					char[] ch2 = word_completion.toCharArray();
					for (int i = 0; i < ch1.length; i++) {
						if (guess.equals("" + ch1[i])) {
							ch2[i] = ch1[i];
						}
					}
					word_completion = String.valueOf(ch2);

					if (!word_completion.contains(symbol)) {
						guessed = true;
					}
				}
			}

			else if (guess.length() == word.length() && guess.matches("[a-zA-Z]+")) {
				if (guessed_letters.contains(guess)) {
					System.out.println("You already guessed the letter " + guess + ". Please try again!");
				} else if (!guess.equals(word)) {
					System.out.println(guess + " is not the WORD. Please try again!");
					tries = tries - 1;
					guessed_letters.add(guess);
				} else {
					word_completion = word;
					guessed = true;
				}
			}

			else {
				System.out.println("Not a valid guess. Please try again!");
			}

			display_Hangman(tries);
			System.out.println("Your word (" + word.length() + " letters): " + word_completion);
			System.out.println();

		}

		if (guessed) {
			System.out.println("Congrats! You guessed the WORD. You win!");
		} else {
			System.out.println("Sorry, you have ran out of tries. The WORD was " + word + ". Maybe next time!");
		}

	}

	public static void display_Hangman(int count) {

		if (count == 6) {
			System.out.println(ANSI_GREEN + "   _______________   ");
			System.out.println(ANSI_RED + "   |   ");
			System.out.println(ANSI_RED + "   |   ");
			System.out.println(ANSI_RED + "   |   ");
			System.out.println(ANSI_RED + "   |   ");
			System.out.println(ANSI_RED + "   |   ");
			System.out.println(ANSI_RED + "   |   ");
			System.out.println(ANSI_RED + "   |   ");
			System.out.println(ANSI_GREEN + "  _|_" + ANSI_RESET);
			System.out.println();
			System.out.println("Chances available: 6");
		}

		if (count == 5) {
			System.out.println("");
			System.out.println(ANSI_GREEN + "   _______________   ");
			System.out.println(ANSI_RED + "   |      " + ANSI_CYAN + "      |  " + ANSI_RESET);
			System.out.println(ANSI_RED + "   |   ");
			System.out.println(ANSI_RED + "   |   ");
			System.out.println(ANSI_RED + "   |   ");
			System.out.println(ANSI_RED + "   |   ");
			System.out.println(ANSI_RED + "   |   ");
			System.out.println(ANSI_RED + "   |   ");
			System.out.println(ANSI_GREEN + "  _|_" + ANSI_RESET);
			System.out.println();
			System.out.println("Chances available: 5");
		}

		if (count == 4) {
			System.out.println("");
			System.out.println(ANSI_GREEN + "   _______________   ");
			System.out.println(ANSI_RED + "   |      " + ANSI_CYAN + "     _|_ " + ANSI_RESET);
			System.out.println(ANSI_RED + "   |      " + ANSI_CYAN + "   |     |  " + ANSI_RESET);
			System.out.println(ANSI_RED + "   |      " + ANSI_CYAN + "   |_ _ _|" + ANSI_RESET);
			System.out.println(ANSI_RED + "   |   ");
			System.out.println(ANSI_RED + "   |   ");
			System.out.println(ANSI_RED + "   |   ");
			System.out.println(ANSI_RED + "   |   ");
			System.out.println(ANSI_GREEN + "  _|_" + ANSI_RESET);
			System.out.println();
			System.out.println("Chances available: 4");
		}

		if (count == 3) {
			System.out.println("");
			System.out.println(ANSI_GREEN + "   _______________   ");
			System.out.println(ANSI_RED + "   |    " + ANSI_CYAN + "       _|_  " + ANSI_RESET);
			System.out.println(ANSI_RED + "   |    " + ANSI_CYAN + "     |     |  " + ANSI_RESET);
			System.out.println(ANSI_RED + "   |    " + ANSI_CYAN + "     |_ _ _|" + ANSI_RESET);
			System.out.println(ANSI_RED + "   |    " + ANSI_CYAN + "        |" + ANSI_RESET);
			System.out.println(ANSI_RED + "   |    " + ANSI_CYAN + "        |" + ANSI_RESET);
			System.out.println(ANSI_RED + "   |    " + ANSI_CYAN + "        |" + ANSI_RESET);
			System.out.println(ANSI_RED + "   |   ");
			System.out.println(ANSI_GREEN + "  _|_" + ANSI_RESET);
			System.out.println();
			System.out.println("Chances available: 3");
		}

		if (count == 2) {
			System.out.println("");
			System.out.println(ANSI_GREEN + "   _______________   ");
			System.out.println(ANSI_RED + "   |      " + ANSI_CYAN + "     _|_  " + ANSI_RESET);
			System.out.println(ANSI_RED + "   |      " + ANSI_CYAN + "   |     |  " + ANSI_RESET);
			System.out.println(ANSI_RED + "   |      " + ANSI_CYAN + "   |_ _ _|" + ANSI_RESET);
			System.out.println(ANSI_RED + "   |      " + ANSI_CYAN + "    --|--" + ANSI_RESET);
			System.out.println(ANSI_RED + "   |      " + ANSI_CYAN + "   || | ||   " + ANSI_RESET);
			System.out.println(ANSI_RED + "   |      " + ANSI_CYAN + "      |" + ANSI_RESET);
			System.out.println(ANSI_RED + "   |   ");
			System.out.println(ANSI_GREEN + "  _|_" + ANSI_RESET);
			System.out.println();
			System.out.println("Chances available: 2");
		}

		if (count == 1) {
			System.out.println("");
			System.out.println(ANSI_GREEN + "   _______________   ");
			System.out.println(ANSI_RED + "   |      " + ANSI_CYAN + "     _|_  " + ANSI_RESET);
			System.out.println(ANSI_RED + "   |      " + ANSI_CYAN + "   |     |  " + ANSI_RESET);
			System.out.println(ANSI_RED + "   |      " + ANSI_CYAN + "   |_ _ _|" + ANSI_RESET);
			System.out.println(ANSI_RED + "   |      " + ANSI_CYAN + "    --|--" + ANSI_RESET);
			System.out.println(ANSI_RED + "   |      " + ANSI_CYAN + "   || | ||   " + ANSI_RESET);
			System.out.println(ANSI_RED + "   |      " + ANSI_CYAN + "      |" + ANSI_RESET);
			System.out.println(ANSI_RED + "   |      " + ANSI_CYAN + "  ' /' '\\' " + ANSI_RESET);
			System.out.println(ANSI_GREEN + "  _|_" + ANSI_RESET);
			System.out.println();
			System.out.println("Chances available: 1");
		}

		if (count == 0) {
			System.out.println("");
			System.out.println(ANSI_GREEN + "   _______________   ");
			System.out.println(ANSI_RED + "   |      " + ANSI_CYAN + "     _|_ " + ANSI_RESET);
			System.out.println(ANSI_RED + "   |      " + ANSI_CYAN + "   | . . |  " + ANSI_RESET);
			System.out.println(ANSI_RED + "   |      " + ANSI_CYAN + "   |_ - _|" + ANSI_RESET);
			System.out.println(ANSI_RED + "   |      " + ANSI_CYAN + "    --|--" + ANSI_RESET);
			System.out.println(ANSI_RED + "   |      " + ANSI_CYAN + "   || | ||   " + ANSI_RESET);
			System.out.println(ANSI_RED + "   |      " + ANSI_CYAN + "      |" + ANSI_RESET);
			System.out.println(ANSI_RED + "   |      " + ANSI_CYAN + "   '/' '\\' " + ANSI_RESET);
			System.out.println(ANSI_GREEN + "  _|_" + ANSI_RESET);
			System.out.println();
			System.out.println("Chances available: 0");
		}
	}

	public static void main(String[] args) {
		String word = get_word();
		play(word);
	}

}
